File: java/sdk/src/main/java/network/t0/sdk/crypto/Keccak256.java
package network.t0.sdk.crypto;

import org.bouncycastle.jcajce.provider.digest.Keccak;

/**
 * Keccak-256 hash function implementation.
 *
 * <p>This uses the original Keccak algorithm (before NIST standardization as SHA-3),
 * which is the same algorithm used by Ethereum for signing.
 *
 * <p>Note: SHA3-256 (NIST SHA-3) uses different padding than Keccak-256.
 * Ethereum and this SDK use the original Keccak-256 ("Legacy Keccak").
 *
 * <p><b>Thread Safety:</b> This class is thread-safe. All methods are stateless and can be
 * called concurrently from multiple threads without synchronization.
 */
public final class Keccak256 {

    private Keccak256() {
        // Utility class
    }

    /**
     * Computes the Keccak-256 hash of the input data.
     *
     * @param data the input data to hash
     * @return 32-byte Keccak-256 hash
     */
    public static byte[] hash(byte[] data) {
        if (data == null) {
            throw new IllegalArgumentException("data must not be null");
        }

        Keccak.Digest256 digest = new Keccak.Digest256();
        digest.update(data);
        return digest.digest();
    }

    /**
     * Computes the Keccak-256 hash of multiple byte arrays concatenated.
     *
     * @param parts the byte arrays to concatenate and hash
     * @return 32-byte Keccak-256 hash
     */
    public static byte[] hash(byte[]... parts) {
        Keccak.Digest256 digest = new Keccak.Digest256();
        for (byte[] part : parts) {
            if (part != null) {
                digest.update(part);
            }
        }
        return digest.digest();
    }
}

File: csharp/sdk/T0.ProviderSdk/Crypto/Keccak256.cs
using Org.BouncyCastle.Crypto.Digests;

namespace T0.ProviderSdk.Crypto;

/// <summary>
/// Keccak-256 hash function (legacy Keccak, NOT NIST SHA-3).
/// This is the same algorithm used by Ethereum.
/// Thread-safe: all methods are stateless.
/// </summary>
public static class Keccak256
{
    /// <summary>
    /// Computes the Keccak-256 hash of the input data.
    /// </summary>
    /// <returns>32-byte Keccak-256 hash.</returns>
    public static byte[] Hash(byte[] data)
    {
        ArgumentNullException.ThrowIfNull(data);

        var digest = new KeccakDigest(256);
        digest.BlockUpdate(data, 0, data.Length);
        var result = new byte[32];
        digest.DoFinal(result, 0);
        return result;
    }

    /// <summary>
    /// Computes the Keccak-256 hash of multiple byte arrays concatenated.
    /// </summary>
    public static byte[] Hash(params byte[][] parts)
    {
        var digest = new KeccakDigest(256);
        foreach (var part in parts)
        {
            if (part is not null)
                digest.BlockUpdate(part, 0, part.Length);
        }

        var result = new byte[32];
        digest.DoFinal(result, 0);
        return result;
    }
}
